asf_heat_pump_suitability/pipeline/prepare_features/grid_capacity.py

```diff
@@ -0,0 +1,758 @@
+        return np.nan
+
+
+def process_shapefile(shapefile_directory: str) -> gpd.GeoDataFrame:
+    """
+    Process a shapefile from S3 and return it as a GeoDataFrame.
+
+    Args:
+        shapefile_directory: S3 directory containing the shapefile.
+
+    Returns:
+        GeoDataFrame containing the processed shapefile data.
+    """
+    bucket_name = "asf-heat-pump-suitability"
+    temp_dir = get_shapefile_from_s3(bucket_name, shapefile_directory)
+
+    try:
+        shp_file = next(f for f in os.listdir(temp_dir) if f.endswith(".shp"))
+        shp_path = os.path.join(temp_dir, shp_file)
+        gdf = gpd.read_file(shp_path)
+        gdf = gdf.to_crs(epsg=27700)  # Convert to British National Grid
+        return gdf
+    finally:
+        shutil.rmtree(temp_dir)
+
+
+def generate_enw_df() -> gpd.GeoDataFrame:
+    """
+    Generate a GeoDataFrame for Electricity North West (ENW) substations.
+
+    Returns:
+        GeoDataFrame with ENW substation data.
+    """
+    enw_demand = load_s3_data(
+        "asf-heat-pump-suitability",
+        "source_data/grid_operators/ENW/dfes-2023-primary-data0.parquet",
+    )
+    enw_substations = load_s3_data(
+        "asf-heat-pump-suitability",
+        "source_data/grid_operators/ENW/ndp-pry-bsp-headroom.parquet",
+    )
+    enw_shape = load_s3_data(
+        "asf-heat-pump-suitability",
+        "source_data/grid_operators/ENW/ndp-pry-voronoi.parquet",
+    )
+
+    # Filter data for primary substations, current year, and best view scenario
+    enw_substations = enw_substations[
+        (enw_substations["substation_type"] == "PRIMARY")
+        & (enw_substations["year"] == "2024")
+        & (enw_substations["scenario"] == "1 - BEST VIEW")
+        & (enw_substations["status"] == "FIRM")
+    ]
+    enw_demand = enw_demand[
+        (enw_demand.group == "PRIMARY")
+        & (enw_demand.year == "2024")
+        & (enw_demand.scenario == "Best View")
+    ]
+
+    # Merge substation and demand data
+    enw = pd.merge(
+        enw_substations[["substation", "headroom_mva", "geopoint"]],
+        enw_demand[["number", "maximum_demand_mva_per_primary_substation"]],
+        left_on="substation",
+        right_on="number",
+    )
+
+    # Calculate firm capacity and add operator information
+    enw["firm_capacity_mva"] = (
+        enw["maximum_demand_mva_per_primary_substation"] + enw["headroom_mva"]
+    )
+    enw["operator"] = "ENW"
+    enw["geopoint"] = enw["geopoint"].apply(_parse_binary_geometry)
+
+    # Create GeoDataFrame and perform spatial join with shape data
+    enw_gdf = gpd.GeoDataFrame(enw, geometry="geopoint", crs=CRS)
+    enw_shape["geo_shape"] = enw_shape["geo_shape"].apply(_parse_binary_geometry)
+    enw_shape["geo_point_2d"] = enw_shape["geo_point_2d"].apply(_parse_binary_geometry)
+    enw_shape_gdf = gpd.GeoDataFrame(enw_shape, geometry="geo_shape", crs=CRS)
+    enw_df = gpd.sjoin(enw_gdf, enw_shape_gdf, how="right", predicate="within")
```

Comment:
```suggestion
    enw_df = gpd.sjoin(enw_gdf, enw_shape_gdf, how="right", predicate="intersects")
```
I have been using the `intersects` predicate because it returns `True` if the boundary or interior of the object intersect in any way with those of the other, so it will match points which are on the boundaries as well as within the polygons. https://shapely.readthedocs.io/en/latest/manual.html#object.intersects

I see you used intersects in the function below so was there a specific reason to use `within` here?